adapters/repos/db/lsmkv/migrator_filterable_to_rangeable.go

```diff
@@ -0,0 +1,751 @@
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package lsmkv
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/helpers"
+	"github.com/weaviate/weaviate/adapters/repos/db/lsmkv/segmentindex"
+	"github.com/weaviate/weaviate/adapters/repos/db/roaringsetrange"
+	"github.com/weaviate/weaviate/entities/errorcompounder"
+	"github.com/weaviate/weaviate/entities/errors"
+)
+
+const (
+	migrationDir = "migration"
+
+	migrationFileInitedGlob   = "migration.inited.*.mig"
+	migrationFileInitedRegex  = "migration.inited.([0-9]{18,})\\.mig"
+	migrationFileInitedFormat = "migration.inited.%d.mig"
+	migrationFileTodo         = "migration.todo.mig"
+	migrationFileCompleted    = "migration.completed.mig"
+
+	segmentFileTodoGlob   = "segment-*.db.todo.mig"
+	segmentFileTodoRegex  = "segment-[0-9]{18,}\\.db\\.todo\\.mig"
+	segmentFileTodoSuffix = ".todo.mig"
+
+	segmentFileTempGlob        = "segment-*.db.*.temp"
+	segmentFileTempRegex       = "segment-[0-9]{18,}\\.db\\.[0-9]{4,}(-[0-9]{4,})?\\.temp"
+	segmentFileTempFormat      = "%s.%04d.temp"
+	segmentFileTempMergeFormat = "%s.%04d-%04d.temp"
+
+	segmentFileDbGlob  = "segment-*.db"
+	segmentFileDbRegex = "segment-([0-9]{18,})\\.db"
+)
+
+var (
+	regexMigrationFileInited = regexp.MustCompile(migrationFileInitedRegex)
+	regexSegmentFileTodo     = regexp.MustCompile(segmentFileTodoRegex)
+	regexSegmentFileTemp     = regexp.MustCompile(segmentFileTempRegex)
+	regexSegmentFileDb       = regexp.MustCompile(segmentFileDbRegex)
+)
+
+type MigratorFilterableToRangeable struct {
+	propName      string
+	logger        logrus.FieldLogger
+	routinesLimit int
+	srcBucket     *Bucket
+	dstBucket     *Bucket
+}
+
+func NewMigratorFilterableToRangeable(store *Store, propName string, routinesLimit int,
+) (*MigratorFilterableToRangeable, error) {
+	srcBucket := store.Bucket(helpers.BucketFromPropNameLSM(propName))
+	if srcBucket == nil {
+		return nil, fmt.Errorf("rangeable migration %q: filterable bucket not found", propName)
+	}
+	dstBucket := store.Bucket(helpers.BucketRangeableFromPropNameLSM(propName))
+	if dstBucket == nil {
+		return nil, fmt.Errorf("rangeable migration %q: rangeable bucket not found", propName)
+	}
+
+	return &MigratorFilterableToRangeable{
+		propName: propName,
+		logger: store.logger.
+			WithField("action", "migrate_filterable_to_rangeable").
+			WithField("property", propName),
+		routinesLimit: routinesLimit,
+		srcBucket:     srcBucket,
+		dstBucket:     dstBucket,
+	}, nil
+}
+
+func (m *MigratorFilterableToRangeable) Migrate(ctx context.Context) error {
```

Comment:
Is the migration process atomic, in that if the migration fails, is it recoverable, or could it potentially lead to corrupted indices?

Reply:
I've tried to make it as safe as possible to be recoverable, so in case process is interrupted all unfinished segments should be automatically removed and migrated from the scratch. i will describe the process in PR description

Reply:
migration script moved to separate PR https://github.com/weaviate/weaviate/pull/5227